sos/plugins/__init__.py

```diff
@@ -150,12 +150,40 @@ def __init__(self, commons):
         self.soslog = self.commons['soslog'] if 'soslog' in self.commons \
             else logging.getLogger('sos')
 
+        # add the 'timeout' plugin option automatically
+        self.option_list.append(('timeout', 'timeout in seconds for plugin',
+                                 'fast', -1))
+
         # get the option list into a dictionary
         for opt in self.option_list:
             self.opt_names.append(opt[0])
             self.opt_parms.append({'desc': opt[1], 'speed': opt[2],
                                    'enabled': opt[3]})
 
+    @property
+    def timeout(self):
+        '''Returns either the default plugin timeout value, the value as
+        provided on the commandline via -k plugin.timeout=value, or the value
+        of the global --plugin-timeout option.
+        '''
+        _timeout = None
+        try:
+            opt_timeout = self.get_option('plugin_timeout')
+            own_timeout = int(self.get_option('timeout'))
+            if opt_timeout is None:
+                _timeout = own_timeout
+            elif opt_timeout is not None and own_timeout == -1:
+                _timeout = int(opt_timeout)
+            elif opt_timeout is not None and own_timeout > -1:
+                _timeout = own_timeout
+            else:
+                return None
+        except ValueError:
+            return self.plugin_timeout  # Default to known safe value
+        if _timeout is not None and _timeout > -1:
+            return _timeout
+        return self.plugin_timeout
+
```

Comment:
I thought the `elif` are wrong but I am unable to find particular counter-example.

Can't something be simplified if we allow just numerical values for the options? Like `./sosreport --log-size=a1` fails with `sosreport: error: argument --log-size: invalid int value: 'a1'`?

Reply:
Havent realized that complexity behind, what can easily justify that code, then.

Reply:
Yeah, that kinda sucks but as @TurboTurtle explained it's a consequence of how the Python bits deal with certain special values. All that said, it's really not too bad - we have far more icky blocks in the rest of the codebase, and the logic is quite easy to follow if you understand the intent of the precedence.